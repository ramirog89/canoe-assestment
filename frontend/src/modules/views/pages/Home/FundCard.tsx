import * as React from 'react';

import Dialog from '../../common/Dialog';
import FundForm from './FundForm';

import { FundModel } from '../../../../models';
import { useFunds } from '../../../hooks/useFunds';

interface IFundCardProps {
  id: number;
  items: FundModel.IFund[];
  type: 'create' | 'edit';
  isOpen: boolean;
  onClose: () => void;
  onSubmit: () => void;
}

const FundCard = ({ id, items, type, isOpen, onClose, onSubmit }: IFundCardProps) => {
  const fundManager = useFunds();

  const [state, setState] = React.useState<FundModel.IFundRequest>(FundModel.getFundRequestFallback());

  const onChangeHandler = (event: any) => {
    setState({
      ...state,
      [event.target.name]: event.target.value
    })
  }

  const onConfirm = () => {
    if (type === 'create') {
      fundManager.create.mutate(state);
    } else {
      fundManager.update.mutate({ id, payload: state });
    }
  }

  const isLoading = fundManager.create.isPending || fundManager.update.isPending;

  React.useEffect(() => {
    if (fundManager.create.isSuccess || fundManager.update.isSuccess) {
      setState(FundModel.getFundRequestFallback());
      onSubmit();
    }
  }, [fundManager.create.isSuccess, fundManager.update.isSuccess]); // eslint-disable-line

  React.useEffect(() => {
    if (id && items) {
      const selectedFund = items.find((fund) => fund.id === id) as any;
      setState({
        name: selectedFund?.name,
        start_year: selectedFund?.start_year,
        alias: selectedFund?.fundalias_set.map((a: any) => a.alias),
        manager: selectedFund.manager.id,
      });
    }
  }, [items, id]); // eslint-disable-line

  return (
    <Dialog
      isOpen={isOpen}
      isLoading={isLoading}
      title={FundModel.modalMap[type]?.title}
      closeLabel={FundModel.modalMap[type]?.closeLabel}
      submitLabel={FundModel.modalMap[type]?.submitLabel}
      onClose={onClose}
      onConfirm={onConfirm}
      render={() => (
        <FundForm
          fund={state}
          onChange={onChangeHandler}
        />
      )}
    />
  );
};

export default FundCard;
