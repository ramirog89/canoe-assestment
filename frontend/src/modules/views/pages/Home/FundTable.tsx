import * as React from 'react';

import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import IconButton from '@mui/material/IconButton';
import Tooltip from '@mui/material/Tooltip';
import DeleteIcon from '@mui/icons-material/Delete';
import EditIcon from '@mui/icons-material/Edit';
import TablePagination from '@mui/material/TablePagination';

import { FundModel } from '../../../../models';

interface IFundTableProps {
  page: number;
  rowsPerPage: number;
  totalItems: number;
  items: FundModel.IFund[];
  onChangePage: (e: any, page: number) => void;
  onChangeRow: (e: any) => void;
  onOpenEditModal: (id: number) => void;
  onOpenDeleteModal: (id: number) => void;
}

const FundTable = ({
  page,
  rowsPerPage,
  totalItems,
  items,
  onChangePage,
  onChangeRow,
  onOpenEditModal,
  onOpenDeleteModal
}: IFundTableProps) => {
  return (
    <React.Fragment>
      <Table size="small">
        <TableHead>
          <TableRow>
            <TableCell>Id</TableCell>
            <TableCell>Name</TableCell>
            <TableCell>Start Year</TableCell>
            <TableCell>Manager</TableCell>
            <TableCell>Created At</TableCell>
            <TableCell></TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {items.map((fund) => (
            <TableRow key={fund.id}>
              <TableCell>{fund.id}</TableCell>
              <TableCell>{fund.name}</TableCell>
              <TableCell>{fund.start_year}</TableCell>
              <TableCell>{fund.manager.name}</TableCell>
              <TableCell>{fund.created_at.toString()}</TableCell>
              <TableCell align="right">
                <Tooltip arrow placement="left" title="Edit">
                  <IconButton onClick={() => onOpenEditModal(fund.id)}>
                    <EditIcon />
                  </IconButton>
                </Tooltip>
                <Tooltip arrow placement="right" title="Delete">
                  <IconButton color="error" onClick={() => onOpenDeleteModal(fund.id)}>
                    <DeleteIcon />
                  </IconButton>
                </Tooltip>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
      <TablePagination
        component="div"
        rowsPerPageOptions={[5, 10, 25]}
        page={page - 1}
        rowsPerPage={rowsPerPage}
        count={totalItems}
        onPageChange={onChangePage}
        onRowsPerPageChange={onChangeRow}
      />
    </React.Fragment>
  )
}

export default FundTable;
