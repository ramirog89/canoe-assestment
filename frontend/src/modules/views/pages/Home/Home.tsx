import * as React from 'react';

import Typography from '@mui/material/Typography';
import Grid from '@mui/material/Grid';
import Paper from '@mui/material/Paper';
import Button from '@mui/material/Button';
import MenuItem from '@mui/material/MenuItem';
import Select from '@mui/material/Select';
import MuiAlert from '@mui/material/Alert';

import Layout from '../../common/Layout';
import Dialog from '../../common/Dialog';
import FundCard from './components/FundCard';
import FundTable from './components/FundTable';

import { useFunds } from '../../../hooks/useFunds';
import { usePaginatedFunds, FundFilter } from '../../../hooks/usePaginatedFunds';

const Home = () => {
  const fundManager = usePaginatedFunds();
  const { create, remove, update } = useFunds();

  const [time, setTime] = React.useState(0);
  const [modal, setModal] = React.useState<any>({ isOpen: false, type: null, id: null });
  const [deleteModal, setDeleteModal] = React.useState<any>({ isOpen: false, id: null });

  const onOpenEditModal = (id: number) => {
    setModal({ isOpen: true, type: 'edit', id });
  }
  
  const onOpenCreateModal = (e: any) => {
    setModal({ isOpen: true, type: 'create', id: null });
  }

  const onOpenDeleteModal = (id: number) => {
    setDeleteModal({ isOpen: true, id });
  }

  const onCloseModal = () => {
    setModal({ ...modal, isOpen: false, id: null });
  }

  const onCloseDeleteModal = () => {
    setDeleteModal({ isOpen: false });
  }

  const onDeleteConfirm = () => {
    remove.mutate({ id: deleteModal.id });
  }

  const onSubmitFund = () => {
    fundManager.funds.refetch();
    create.reset();
    update.reset();
    onCloseModal();
  }

  React.useEffect(() => {
    if (remove.isSuccess) {
      fundManager.funds.refetch();
      remove.reset();
      onCloseDeleteModal();
    }
  }, [remove.isSuccess]); // eslint-disable-line

  React.useEffect(() => {
    const timer = setTimeout(() => {
      setTime(Date.now())
    }, 5000);
    if (!deleteModal.isOpen && !modal.isOpen) {
      fundManager.funds.refetch();
    }
    return () => clearTimeout(timer);
  }, [time, deleteModal.isOpen, modal.isOpen]); // eslint-disable-line

  return (
    <Layout>
      <Grid container spacing={3}>
        <Grid item xs={12}>
          <MuiAlert elevation={6} variant="filled" severity='info'>
            List is being refreshed every 30 secs to show new generated Records
          </MuiAlert>
          <Paper sx={{ p: 2, display: 'flex', flexDirection: 'column', mt: '10px' }}>
            <div style={{
              display: 'flex',
              flexDirection: 'row',
              justifyContent: 'space-between',
              padding: '10px 0'
            }}>
              <div style={{
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'space-between',
                alignItems: 'center'
              }}>
                <Typography component="h2" variant="h6" color="primary">
                  Funds
                </Typography>
                <Select
                  label="Manager"
                  name="manager"
                  size="small"
                  value={fundManager.filter}
                  style={{ marginLeft: '20px' }}
                  onChange={(event) => fundManager.onChangeFilter(event.target.value)}
                >
                  <MenuItem value={FundFilter.ALL}>{FundFilter.ALL}</MenuItem>
                  <MenuItem value={FundFilter.DUPLICATES}>{FundFilter.DUPLICATES}</MenuItem>
                </Select>
              </div>
              <div>
                <Button color="primary" onClick={onOpenCreateModal} variant="contained">
                  Create
                </Button>
              </div>
            </div>
            {!fundManager.funds.isLoading && fundManager.funds?.data?.total === 0 ? (
              <div>
                Empty Records
              </div>
            ) : <></>}
            {fundManager.funds?.data?.total && fundManager.funds?.data?.total >= 0 ? (
              <FundTable
                isLoading={fundManager.funds.isLoading}
                page={fundManager.page}
                rowsPerPage={fundManager.rowsPerPage}
                items={fundManager.funds.data?.items || []}
                totalItems={fundManager.funds?.data?.total || 0}
                onChangePage={fundManager.onChangePage}
                onChangeRow={fundManager.onChangeRowsPerPage}
                onOpenEditModal={onOpenEditModal}
                onOpenDeleteModal={onOpenDeleteModal}
              />
              ) : <></>}
          </Paper>
        </Grid>
      </Grid>
      <FundCard
        id={modal.id as any}
        items={fundManager.funds.data?.items || []}
        type={modal.type}
        isOpen={modal.isOpen}
        onClose={onCloseModal}
        onSubmit={onSubmitFund}
      />
      <Dialog
        title="Delete Fund"
        isLoading={false}
        isOpen={deleteModal.isOpen}
        onClose={onCloseDeleteModal}
        onConfirm={onDeleteConfirm}
        closeLabel='Cancel'
        submitLabel='Confirm'
        render={() => (
          <>
            Are you sure you want to delete Fund #{deleteModal.id}?
          </>
        )}
      />
    </Layout>
  );
};

export default Home;
