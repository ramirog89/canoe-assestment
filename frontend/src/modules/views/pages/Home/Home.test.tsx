import { render, RenderResult } from "../../../../test/render";

import Home from './Home';

describe('Home', () => {
  let wrapper: RenderResult;

  it('should render', () => {
    wrapper = render(<Home />);
    expect(wrapper.container).toMatchSnapshot();
  });
});
