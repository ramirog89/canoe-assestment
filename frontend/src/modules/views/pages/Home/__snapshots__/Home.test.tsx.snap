// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render 1`] = `
<div>
  <div
    class="MuiBox-root css-k008qs"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-fjo2tc-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h6 css-1qp5wg9-MuiTypography-root"
        >
          Dashboard
        </h1>
      </div>
    </header>
    <main
      class="MuiBox-root css-1w1gdci"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      />
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1f2kw6w-MuiContainer-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 MuiAlert-root MuiAlert-filledInfo MuiAlert-filled css-1othqhp-MuiPaper-root-MuiAlert-root"
              role="alert"
            >
              <div
                class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
                  data-testid="InfoOutlinedIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11,9H13V7H11M12,20C7.59,20 4,16.41 4,12C4,7.59 7.59,4 12,4C16.41,4 20,7.59 20, 12C20,16.41 16.41,20 12,20M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10, 10 0 0,0 12,2M11,17H13V11H11V17Z"
                  />
                </svg>
              </div>
              <div
                class="MuiAlert-message css-1pxa9xg-MuiAlert-message"
              >
                List is being refreshed every 30 secs to show new generated Records
              </div>
            </div>
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-3yxrt8-MuiPaper-root"
            >
              <div
                style="display: flex; flex-direction: row; justify-content: space-between; padding: 10px 0px;"
              >
                <div
                  style="display: flex; flex-direction: row; justify-content: space-between; align-items: center;"
                >
                  <h2
                    class="MuiTypography-root MuiTypography-h6 css-1bbxbns-MuiTypography-root"
                  >
                    Funds
                  </h2>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall  css-r6dbjx-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
                    style="margin-left: 20px;"
                  >
                    <div
                      aria-controls=":r0:"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-labelledby="mui-component-select-manager"
                      class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                      id="mui-component-select-manager"
                      role="combobox"
                      tabindex="0"
                    >
                      all
                    </div>
                    <input
                      aria-hidden="true"
                      aria-invalid="false"
                      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                      name="manager"
                      tabindex="-1"
                      value="all"
                    />
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-yjsfm1"
                      >
                        <span>
                          Manager
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
                <div>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-nqi81f-MuiButtonBase-root-MuiButton-root"
                    tabindex="0"
                    type="button"
                  >
                    Create
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
  </div>
</div>
`;
